File: leetcode/easy/20-valid-parentheses/solution.js
function isValid(s) {
  if (s.length % 2 !== 0 || s[0] === ']' || s[0] === '}' || s[0] === ')') {
    return false;
  }

  const stack = [];

  for (let i = 0; i < s.length; i++) {
    if (stack.length === 0) {
      stack.push(s[i]);
      continue;
    }
    if (s[i] === '[' || s[i] === '{' || s[i] === '(') {
      stack.push(s[i]);
      continue;
    }
    if (s[i] === ')' && stack[stack.length - 1] === '(') {
      stack.pop();
      continue;
    }
    if (s[i] === '}' && stack[stack.length - 1] === '{') {
      stack.pop();
      continue;
    }
    if (s[i] === ']' && stack[stack.length - 1] === '[') {
      stack.pop();
      continue;
    }
    return false;
  }

  return stack.length === 0;
}

File: leetcode/easy/20-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:

      open_bracket = '[{('
      close_bracket = ']})'

      if len(s) % 2 != 0 or s[0] in close_bracket:
        return False

      stack = []

      for bracket in s:
        if len(stack) == 0 or bracket in open_bracket:
          stack.append(bracket)
          continue
        if bracket == close_bracket[0] and stack[-1] == open_bracket[0]:
          stack.pop()
          continue
        if bracket == close_bracket[1] and stack[-1] == open_bracket[1]:  
          stack.pop()
          continue
        if bracket == close_bracket[2] and stack[-1] == open_bracket[2]:  
          stack.pop()
          continue
        return False
      
      return len(stack) == 0

